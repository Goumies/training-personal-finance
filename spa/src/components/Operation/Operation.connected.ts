import { State, Updater, connect } from "../../store";
import { OperationComponent, OwnProps, StateProps, UpdateProps } from "./Operation";
import { Operation as IOperation } from "../../interfaces/Operation";

function computeStateProps({ operations }: State, { id }: OwnProps): StateProps {
  const index: number = operations.findIndex(({ id: currentId }: IOperation) => currentId === id);
  if (index === -1) {
    throw new Error(`No operation matches the following id: ${id}`);
  }
  const operation: IOperation = operations[index];
  return { operation };
}

function computeUpdateProps(updateState: (updater: Updater) => void, { id }: OwnProps): UpdateProps {
  const onDelete = () => {
    const updater: Updater = (state: State) => {
      const operations: IOperation[] = [...state.operations];
      const index: number = operations.findIndex(({ id: currentId }: IOperation) => currentId === id);
      if (index === -1) {
        return state;
      }
      operations.splice(index, 1);
      return {
        ...state,
        operations
      };
    };
    updateState(updater);
  };
  return { onDelete };
}

export const Operation = connect<OwnProps, StateProps, UpdateProps>(
  OperationComponent,
  { computeStateProps, computeUpdateProps }
);
