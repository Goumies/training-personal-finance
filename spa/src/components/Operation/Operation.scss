.operation {
  position: relative;

  .actions-container {
    position: absolute;
    box-sizing: border-box;
    width: 100px;
    height: 100%;
    padding: 0 20px;
    top: -5px;
    right: -100px;

    .actions-content {
      display: none;
    }
  }

  &:hover {
    .actions-content {
      display: block;
    }
  }
}
