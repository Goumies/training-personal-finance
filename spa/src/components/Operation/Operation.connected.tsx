import React from "react";
import { StoreContext, Store, State, Updater } from "../../store";
import { OperationComponent, OwnProps } from "./Operation";
import { Operation as IOperation } from "../../interfaces/Operation";

export const Operation: React.StatelessComponent<OwnProps> = ({ id }: OwnProps) => (
  <StoreContext.Consumer>
    {(store: Store | undefined) => {
      if (store) {
        const operation: IOperation = getOperation(store.getState(), id);
        const onDelete = () => {
          const updater: Updater = (state: State) => {
            const operations: IOperation[] = [...state.operations];
            const index: number = operations.findIndex(({ id: currentId }: IOperation) => currentId === id);
            if (index === -1) {
              return state;
            }
            operations.splice(index, 1);
            return {
              ...state,
              operations
            };
          };
          store.updateState(updater);
        };
        return <OperationComponent id={id} operation={operation} onDelete={onDelete} />;
      }
    }}
  </StoreContext.Consumer>
);

function getOperation({ operations }: State, id: string): IOperation {
  const index: number = operations.findIndex(({ id: currentId }: IOperation) => currentId === id);
  if (index === -1) {
    throw new Error(`No operation matches the following id: ${id}`);
  }
  return operations[index];
}
