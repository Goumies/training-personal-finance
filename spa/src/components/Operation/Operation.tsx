import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import { FinanceDate } from "../FinanceDate";
import { LabeledDebitOrCredit } from "../LabeledDebitOrCredit";
import { OperationActions } from "../OperationActions/OperationActions";
import "./Operation.scss";

export interface OwnProps {
  id: string;
}

export interface StateProps {
  operation: IOperation;
}

export interface UpdateProps {
  onDelete(): void;
}

export type Props = OwnProps & StateProps & UpdateProps;

export const OperationComponent: React.StatelessComponent<Props> = ({
  operation: { id, date, amount },
  onDelete
}: Props) => {
  return (
    <div className="operation" data-e2e="account-operation">
      <LabeledDebitOrCredit amount={amount} renderLabel={() => <FinanceDate date={date} />} />
      <div className="actions-container">
        <div className="actions-content">
          <OperationActions onDelete={onDelete} />
        </div>
      </div>
    </div>
  );
};
