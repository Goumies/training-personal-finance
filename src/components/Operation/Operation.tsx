import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import { FinanceDate } from "../FinanceDate";
import { DebitOrCredit } from "../DebitOrCredit";
import "./Operation.scss";

interface Props {
  operation: IOperation;
}

export const Operation: React.StatelessComponent<Props> = ({ operation: { date, amount } }: Props) => {
  return (
    <div className="operation">
      <div className="date-container">
        <FinanceDate date={date} />
      </div>
      <div className="amount-container">
        <DebitOrCredit amount={amount} />
      </div>
    </div>
  );
};
