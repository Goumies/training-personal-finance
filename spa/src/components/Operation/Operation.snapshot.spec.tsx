import React from "react";
import renderer from "react-test-renderer";
import { Operation, Props } from "./Operation";

it("renders correctly", () => {
  const props: Props = {
    operation: {
      id: "0",
      date: new Date(2019, 4, 24),
      amount: 10
    }
  };
  const tree = renderer.create(<Operation {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
