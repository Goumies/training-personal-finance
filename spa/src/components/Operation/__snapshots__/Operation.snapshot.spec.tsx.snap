// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="operation"
  data-e2e="account-operation"
>
  <div
    className="labeled-debit-or-credit"
  >
    <div
      className="label-container"
    >
      <span
        className="date"
      >
        2019-5-24
      </span>
    </div>
    <div
      className="amount-container"
    >
      <div
        className="debit-or-credit"
      >
        <span
          className="credit"
        >
          <span
            className="value"
          >
            +10
          </span>
        </span>
        <span
          className="debit"
        />
      </div>
    </div>
  </div>
</div>
`;
